runtime/cudaq/qis/managers/BasicExecutionManager.h

```diff
@@ -107,24 +129,24 @@ class BasicExecutionManager : public cudaq::ExecutionManager {
 
     if (ctx_name == "observe" || ctx_name == "sample" ||
         ctx_name == "extract-state") {
-      for (auto &q : contextQuditIdsForDeletion) {
-        deallocateQudit(q);
-        returnIndex(q);
-      }
+      deallocateQudits(contextQuditIdsForDeletion);
+      for (auto &q : contextQuditIdsForDeletion)
+        returnIndex(q.id);
+
       contextQuditIdsForDeletion.clear();
     }
     executionContext = nullptr;
   }
 
   std::size_t getAvailableIndex(std::size_t quditLevels) override {
     auto new_id = getNextIndex();
-    allocateQudit({quditLevels, new_id});
+    requestedAllocations.emplace_back(quditLevels, new_id);
```

Comment:
I could imagine moving this lazy allocation to the `QubitQIRExecutionManager` instead of doing this at the base class level. I think that would address your point, and would allow future subtypes to decide whether lazy allocation was good for them or not. What do you think?

Reply:
Update made. See if that looks better.